Bugsnag/Helpers/BugsnagKVStore.c

```diff
@@ -0,0 +1,143 @@
+//
+//  BugsnagKVStore.c
+//  Bugsnag
+//
+//  Created by Karl Stenerud on 11.09.20.
+//  Copyright © 2020 Bugsnag Inc. All rights reserved.
+//
+
+#include "BugsnagKVStore.h"
+
+#include <dirent.h>
+#include <errno.h>
+#include <fcntl.h>
+#include <stdio.h>
+#include <string.h>
+#include <sys/stat.h>
+#include <sys/types.h>
+#include <unistd.h>
+
+static DIR* g_currentDir = NULL;
+static int g_currentDirFD = 0;
+
+static int openKeyRead(const char* name) {
+    return openat(g_currentDirFD, name, O_RDONLY, 0);
+}
+
+static int openKeyWrite(const char* name) {
+    return openat(g_currentDirFD, name, O_CREAT | O_RDWR | O_TRUNC, 0600);
+}
+
+void bsgkv_open(const char* path, int* err) {
+    bsgkv_close();
+
+    g_currentDir = opendir(path);
+    if(g_currentDir == NULL) {
+        if(errno == ENOENT) {
+            if(mkdir(path, 0700) != 0) {
+                *err = errno;
+                return;
+            }
+            g_currentDir = opendir(path);
+            if(g_currentDir == NULL) {
+                *err = errno;
+                return;
+            }
+        } else {
+            *err = errno;
+            return;
+        }
+    }
+
+    g_currentDirFD = dirfd(g_currentDir);
+    if(g_currentDirFD < 0) {
+        *err = errno;
+        return;
+    }
+
+    *err = 0;
+}
+
+void bsgkv_close(void) {
+    if(g_currentDirFD != 0) {
+        close(g_currentDirFD);
+        g_currentDirFD = 0;
+    }
+    if(g_currentDir != NULL) {
+        closedir(g_currentDir);
+        g_currentDir = NULL;
+    }
+}
+
+void bsgkv_delete(const char* key, int* err) {
+    if(unlinkat(g_currentDirFD, key, 0) < 0) {
+        if(errno != ENOENT) {
+            *err = errno;
+            return;
+        }
+    }
+    *err = 0;
+}
+
+void bsgkv_setBytes(const char* key, const uint8_t* value, int length, int* err) {
+    int fd = openKeyWrite(key);
+    if(fd < 0) {
+        *err = errno;
+        return;
+    }
+    if(write(fd, value, length) < 0) {
```

Comment:
Shouldn't this check if the number of bytes written is equal to `length`? It seems possible that bytes written could be less than the intended amount.

Reply:
The only situations that would cause less bytes to be written are:
- Error (returns -1)
- The device ran out of space (so we're screwed no matter what and this write won't matter in the grand scheme of things)
- The call got interrupted by a signal handler. 4.2BSD introduced automatic restart, which will cause certain syscalls (such as `write`) to be re-sent automatically. Darwin is based on 4.4BSD so it has this functionality.

But then again, if we wanted to use this in other operating systems, it would make sense to put in the EINTR handler code anyway...